sympy/functions/combinatorial/factorials.py

```diff
@@ -987,11 +987,11 @@ def eval(cls, n, k):
                 # since adding 1 to it cannot give 0
                 return S.Zero
         if all(int_like(x) for x in (n, k)):
-            if (n*k).is_positive and n_k.is_negative:
+            if (n*k).is_positive and n_k.is_negative: # 0 > k > n
                 return S.Zero
-            if n.is_nonnegative and k.is_negative:
+            if n.is_nonnegative and k.is_negative: # n >= 0 > k
                 return S.Zero
-            if n.is_zero and k.is_nonzero:
+            if n.is_nonnegative and n_k.is_negative: # k > n >= 0
```

Comment:
I could prolly combine these two conditions with `k.is_negative or n_k.is_negative` they'd have the same effect

Reply:
Maybe we can get a test into the factorials test suite that fails like the one in concrete. I don't know what the failure is (different form of same answer or wrong code change) but in either case it would be good to have a test closer to home that would fail.

Reply:
Yupp, I'll port some to `test_comb_factorials`

Reply:
Maybe better to revert this change if it caused the test to fail. Which was better, the before or after form?

Reply:
The after form is way more versatile.  
The thing with the before form is, it doesn't explicitly set the result as `0` when `k > n >= 0` instead,
 it lets `_eval` deal with it. This is why the test didn't fail. ( I don't think `_eval` evaluates for all symbolic args )  
The after form immediately sets the result as S.Zero. (even if both args are symbolic)  
This pointed out that there's something fishy going on with the `Sum` test.

Reply:
I agree that the `n.is_zero and k.is_nonzero` is a subset of n.is_nonnegative and k.is_negative or n_k.is_negative` but the sum result for me gives
```python
>>> Sum(binomial(n, k)*factorial(n - k), (k, 0, n)).doit()
E*factorial(n)
```
And that is wrong for `n = 0 which gives (0,0)0! = 1*1 = 1 != E*0!`

Reply:
> @smichr I was pondering why that test failed. and I think it's a bug with Sum.

https://github.com/sympy/sympy/pull/18960#issuecomment-618682907